doc/compiler/optimizer/IntroLoopOptimizations.md

```diff
@@ -0,0 +1,183 @@
+
+### Versioning with Well-behaved Loops with Variant Checks
+```
+for (i=0; i<n; i++) {
+   a[i] = i; // doing a bound check every time because “i" changes
+   }
+```
+This is a well-behaved loop: `n` and array `a` are invariant inside the loop; The index starts out with a value and increments by `1` every time. The range of `i` is known. As long as `i` is less than `n` to begin with, you will increment `i` until it becomes `n`, at which point, the loop exits.
+
+Since the loop is going to run `n` times, instead of checking `i < 0` and `i >= a.length` inside the loop in each iteration, you can check `i < 0` and `i+n >= a.length` in the loop preheader. The two edges/ends of the range for values of `i` are tested: the enter value `i` and the final value `i+n`, with respect to the two conditions. If the upper bound and the lower bound are guaranteed to be fine, the bound does not need to be checked inside the loop every time.
+
+If any of the conditions fail, you go to the slow loop. Somewhere down the line in the slow loop in some iteration, the bound check exception will be raised. That is good enough to send it to the slow loop and run all iterations in the slow loop.
+
+Therefore, with a well-behaved loop, you could use the starting value to do the less-than test and the ending/final value to do the greater-or-equal-to test. This allows you to eliminate the bound check even if the value changes within the loop.
+
+This trick cannot be used if `i` is not changed in a predictable way. For example, `i=foo()` and the bound check on `a[i]` inside the loop cannot be eliminated. The array itself cannot be changing within the loop either. If array `a` points to an array of `100` elements on the first iteration and points to an array of `1` element on the second iteration, the bound check cannot be eliminated inside the loop.  Another example is that if the index to the array is a field: such as `a[o.f]` and there is a call within the loop. That call could change the field `o.f`. In this case, the bound check on `a[o.f]` cannot be eliminated.
+
+### Versioning with Different Checks and Guards
+
+We do not always loop version `NULL` checks. `NULL` check is free on X86. We execute `NULL` check not via an explicit test on `NULL`. There will be no generated code. We optimized away the `NULL` check in the code generator. The first few pages of the address space are protected. If anyone tries to access a value based off `NULL`, say a field with offset `0x25`, they will be accessing address `0x0+0x25` which falls on the first page in memory which is protected. A trap gets raised. VM handles the trap and checks where this trap comes from. It looks up the JIT meta data for which instruction is trapped. Then it raises `NULL` pointer exception with the right line number.
```

Comment:
> `NULL` check is free on X86

Unless the reader is familiar with the concept they may be left wondering what this is called. Perhaps we can say:

> `NULL` check is free on X86 via the resumable trap mechanism.

Reply:
Also a nitpick but this has nothing to do with x86 architecture per say. The architecture certainly allows us to read the first few pages. It is the operating system which will produce a trap. For example on IBM Z on Linux the OS will produce the trap, but on z/OS we totally can read the first page of memory (and we do because the OS stores some important data there).

Reply:
Addressed in 641a90a